### ISSUE TITLE
Is it possible to validate fields only on `blur` and `submit` and not `change` events?

### ISSUE THREAD

**Issue by sergii-vorobei** [react-hook-form/react-hook-form#283]
**Is your feature request related to a problem? Please describe.**
I need to validate fields only on `blur` and `submit` events. Even when I choose `mode = onBlur`, still it validates the field both on `change` and `blur` events. The validation in `change` event is no needed for me. As far as I understand, `mode = onBlur` should validate only on `blur`. Am I right?

**Describe the solution you'd like**
Make validation only on `blur` and `submit` possible.

**Comment by bluebill1049** [comment]
hmmm, we recently made the change on that to introduce `onChange` for `onblur` with revalidate.  maybe we need a setting for that. FYI @stramel what about `reValidateMode`?

**Comment by barrymay** [comment]
@sergii-vorobei if you have an error, do you not want that error to clear until a blur as well?

**Comment by sergii-vorobei** [comment]
@barrymay What I want is that error clears/disappears on `change` and input is revalidated only on `submit` and `blur` event. As far as I see it, just skip field validation on `change` event.

I was able to achieve something similar with `onChange={(e) => clearError('email')}` but it seems to be ugly and causes issues with async validation.

**Comment by barrymay** [comment]
So is the reason for this ask because you're making an external/expensive call on each validation?

**Comment by sergii-vorobei** [comment]
Yes, this too. But the main reason is that I want to display the error only when field looses focus and during the submit event. Not during the time when the user is entering something in the field.

**Comment by barrymay** [comment]
last question, prior to recent releases, was this a problem?

**Comment by bluebill1049** [comment]
@sergii-vorobei can you try version 3.22.3? see if that version solves your problem.

**Comment by sktt** [comment]
Ah! I didn't know we revalidate onChange  in onBlur mode when there's an error. i've implemented exactly the same manually. How does this work in 3.22.3?

**Comment by bluebill1049** [comment]
@sktt previous version is onBlur For both :)

**Comment by bluebill1049** [comment]
@sergii-vorobei can you confirm if `version 3.22.3` works for you? then I can discuss with @barrymay @stramel in terms of the enhancement.

**Comment by barrymay** [comment]
I think I'm getting the problem - the introduction of on change to his is allowing the validator to be called twice immediate, which could lead to perf problems:

I created a Pen to show this:
https://codepen.io/barrymay/pen/WNeXrMv?&editable=true#anon-signup

I'm not quite sure the best way to solve this and https://github.com/react-hook-form/react-hook-form/issues/266#issuecomment-526448226, but I'm sure it's possible

**Comment by bluebill1049** [comment]
thanks, @barrymay I made a PR for this https://github.com/react-hook-form/react-hook-form/pull/290 
this will improve the performance for this lib :)

**Comment by bluebill1049** [comment]
going to close this issue for now. I think the performance issue has been addressed. in terms of revalidating mode until more people request for such feature then we will consider to add it in. Simple API is always the way to go :)

**Comment by acusti** [comment]
first of all, congrats on this library. it’s the first react form library i’ve seen that i really like, and i applaud your efforts, design choices, and philosophy.

for this particular issue, i think the default user experience here still leaves something to be desired. in my opinion, in a regular multi-input form, if a user is still focused on an input, it is rarely helpful to interrupt them to tell them that the value of the input is invalid (one exception i can think of is for choosing a password, where it is useful to provide realtime feedback about issues with the password they are creating as they enter it). as a result, i think that once a user returns to an input after submitting that had an error and begins changing the value or entering a new one, the form should not show an error again for that input until it is validated again (`onSubmit` or `onBlur`, depending on the mode).

the difference is subtle, but here’s a code sandbox illustrating the distinction: https://codesandbox.io/s/react-hook-form-clearing-input-errors-after-submit-ccqzm

both the first and last name fields are required. if you submit an empty form, both will show the validation error. if you then start entering values for the “first name” field, the error will go away as you type. if you happen to clear the field again as you work on it (like if you make a typo and hit delete, or reconsider what you want to enter), the validation error will once again show, then go away again when you start typing again. to me, that experience is inconsistent from the initial experience and therefore what i am expecting, unhelpful, and jarring.

for the last name input, i added an `onChange` prop with a handler that clears the error for that field (if one exists). the handler:
```js
  const clearInputError = useCallback(
    event => {
      if (errors[event.target.name]) {
        clearError(event.target.name);
      }
    },
    [clearError, errors]
  );
```
the input jsx:
```jsx
        <input
          name="lastName"
          onChange={clearInputError}
          ref={registerRequiredInput}
          type="text"
        />
```
as a result, if you happen to clear the input while still focused on it and editing, the error message does not get rendered, like in the initial (pre-submit) UX of filling out the form. to me, this latter behavior is preferable.

luckily, it’s not a hard-to-implement workaround, though it took me a while to figure out how to get what i wanted. however, needing to create a custom handler plus `onChange` to clear the error isn’t ideal, and the rapid double renders of getting the error, then manually clearing the error via the `onChange` handler would be nice to avoid.

**Comment by bluebill1049** [comment]
hey @acusti thanks for your feedback. I did have a thought about this to introduce an option for revalidating. `reValidateMode` (in fact I had a PR for that before) but was dropped due to low demand and you can make a workaround like what you had above. It's also a tradeoff for me when I am introducing more API to the lib, but I am keeping an eye on this issue if more people having similar, I am more than happy to introduce such feature.

**Comment by bluebill1049** [comment]
@acusti I am asking the community for introducing `reValidateMode`. I will keep you updated on such a feature.

**Comment by nandorojo** [comment]
@ilkerkaran that works for me too! I have a question though: how can I set it to 1) **show** an error `onBlur`, and 2) **clear** an error `onChange`? It seems like your implementation only clears the error `onBlur`.

I tried this:

`useForm({ mode: 'onBlur', reValidateMode: 'onChange' })`

This is great for showing an error on blur, but it also only clears the error on blur.

Code sample here: https://codesandbox.io/s/react-hook-form-get-started-forked-vkk9j?file=/src/index.js:0-748
